src/workerd/jsg/util.c++

```diff
@@ -570,29 +571,60 @@ kj::Array<kj::byte> asBytes(v8::Local<v8::ArrayBufferView> arrayBufferView) {
 }
 
 void recursivelyFreeze(v8::Local<v8::Context> context, v8::Local<v8::Value> value) {
-  if (value->IsArray()) {
-    // Optimize array freezing (Array is a subclass of Object, but we can iterate it faster).
-    v8::HandleScope scope(v8::Isolate::GetCurrent());
-    auto arr = value.As<v8::Array>();
-
-    for (auto i: kj::zeroTo(arr->Length())) {
-      recursivelyFreeze(context, check(arr->Get(context, i)));
-    }
+  auto isolate = v8::Isolate::GetCurrent();
+  v8::HandleScope outerScope(isolate);
+  kj::Vector<v8::Local<v8::Value>> queue;
+  queue.reserve(128);
+  queue.add(value);
+
+  while (!queue.empty()) {
+    auto item = queue.back();
+    queue.removeLast();
+
+    if (item->IsArray()) {
+      auto arr = item.As<v8::Array>();
+      uint32_t length = arr->Length();
+
+      // Process array elements in batches to reduce V8 API call overhead
+      constexpr uint32_t BATCH_SIZE = 32;
+      for (uint32_t i = 0; i < length;) {
+        uint32_t batchEnd = kj::min(i + BATCH_SIZE, length);
+        queue.reserve(queue.size() + (batchEnd - i));
+
+        for (; i < batchEnd; ++i) {
+          auto element = check(arr->Get(context, i));
+          if (!element->IsNullOrUndefined()) {
+            queue.add(element);
+          }
+        }
+      }
 
-    check(arr->SetIntegrityLevel(context, v8::IntegrityLevel::kFrozen));
-  } else if (value->IsObject()) {
-    v8::HandleScope scope(v8::Isolate::GetCurrent());
-    auto obj = value.As<v8::Object>();
-    auto names = check(obj->GetPropertyNames(context, v8::KeyCollectionMode::kOwnOnly,
-        v8::ALL_PROPERTIES, v8::IndexFilter::kIncludeIndices));
+      check(arr->SetIntegrityLevel(context, v8::IntegrityLevel::kFrozen));
+    } else if (item->IsObject()) {
+      auto obj = item.As<v8::Object>();
+      auto names = check(obj->GetPropertyNames(context, v8::KeyCollectionMode::kOwnOnly,
+          v8::ALL_PROPERTIES, v8::IndexFilter::kIncludeIndices));
+      if (!names.IsEmpty()) {
+        uint32_t length = names->Length();
+
+        constexpr uint32_t BATCH_SIZE = 16;
+        for (uint32_t i = 0; i < length;) {
+          uint32_t batchEnd = std::min(i + BATCH_SIZE, length);
+          queue.reserve(queue.size() + (batchEnd - i));
+
+          for (; i < batchEnd; ++i) {
+            auto propValue = check(obj->Get(context, check(names->Get(context, i))));
+            if (!propValue->IsNullOrUndefined()) {
+              queue.add(propValue);
```

Comment:
Since we only actually care about `Array` and `Object`, it would likely be more efficient to filter out non-Array, and non-Object values when adding to the queue, that way we can avoid unnecessary additional allocations via the queue reservations.